work_queue/src/work_queue.c

```diff
@@ -4663,33 +4661,43 @@ static void print_large_tasks_warning(struct work_queue *q)
 	rmsummary_delete(largest_unfit_task);
 }
 
+static void trim_factory( struct work_queue *q, struct work_queue_worker *w )
+{
+	struct work_queue_factory_info *f;
+			f = hash_table_lookup(q->factory_table, w->factory_name);
+			if ( f && f->connected_workers > f->max_workers &&
+					itable_size(w->current_tasks) < 1 ) {
+				debug(D_WQ, "Final task received from worker %s, shutting down.", w->hostname);
+				shut_down_worker(q, w);
+			}
+}
+
 static int receive_one_task( struct work_queue *q )
 {
 	struct work_queue_task *t;
-
 	struct work_queue_worker *w;
 	uint64_t taskid;
 
-	itable_firstkey(q->tasks);
-	while( itable_nextkey(q->tasks, &taskid, (void **) &t) ) {
-		if( task_state_is(q, taskid, WORK_QUEUE_TASK_WAITING_RETRIEVAL) ) {
-			w = itable_lookup(q->worker_task_map, taskid);
-			fetch_output_from_worker(q, w, taskid);
-			// Shutdown worker if appropriate.
-			if ( w->factory_name ) {
-				struct work_queue_factory_info *f;
-				f = hash_table_lookup(q->factory_table, w->factory_name);
-				if ( f && f->connected_workers > f->max_workers &&
-						itable_size(w->current_tasks) < 1 ) {
-					debug(D_WQ, "Final task received from worker %s, shutting down.", w->hostname);
-					shut_down_worker(q, w);
-				}
+	if(!(t = q->last_waiting_task)) {
```

Comment:
I'm going to be persnickety here since this is deep in the core of WQ.

Simplify the test at the top to make it as clear as possible:
```
t  = q->last_waiting_task;
if(!t) {
...
}
```